Fabric.Identity.API/scripts/Install-Identity-Utilities.psm1

```diff
@@ -0,0 +1,621 @@
+    $ct = [System.DirectoryServices.AccountManagement.ContextType]::Domain
+    $pc = New-Object System.DirectoryServices.AccountManagement.PrincipalContext -ArgumentList $ct,$credential.GetNetworkCredential().Domain
+    Write-Host "Confirming credentials"
+    $isValid = $pc.ValidateCredentials($credential.GetNetworkCredential().UserName, $credential.GetNetworkCredential().Password, [System.DirectoryServices.AccountManagement.ContextOptions]::Negotiate)
+    if(!$isValid){
+        Write-DosMessage -Level "Error" -Message "Incorrect credentials for $($credential.GetNetworkCredential().UserName)"
+        throw
+    }
+}
+
+function Add-PermissionToPrivateKey([string] $iisUser, [System.Security.Cryptography.X509Certificates.X509Certificate2] $signingCert, [string] $permission){
+    try{
+        $allowRule = New-Object security.accesscontrol.filesystemaccessrule $iisUser, $permission, allow
+        $keyFolder = "c:\programdata\microsoft\crypto\rsa\machinekeys"
+
+        $keyname = $signingCert.privatekey.cspkeycontainerinfo.uniquekeycontainername
+        $keyPath = [io.path]::combine($keyFolder, $keyname)
+
+        if ([io.file]::exists($keyPath))
+        {        
+            $acl = Get-Acl $keyPath
+            $acl.AddAccessRule($allowRule)
+            Set-Acl $keyPath $acl -ErrorAction Stop
+            Write-DosMessage -Level "Information" -Message "The permission '$($permission)' was successfully added to the private key for user '$($iisUser)'"
+        }else{
+            Write-DosMessage -Level "Error" -Message "No key file was found at '$($keyPath)' for '$($signingCert)'. Ensure a valid signing certificate was provided"
+            throw
+        }
+    }catch{
+        Write-DosMessage -Level "Error" -Message "There was an error adding the '$($permission)' permission for the user '$($iisUser)' to the private key. Ensure you selected a certificate that you have read access on the private key. Error $($_.Exception.Message)."
+        throw
+    }
+}
+
+function Get-AppInsightsKey([string] $appInsightsInstrumentationKey, [bool] $quiet){
+    if(!$quiet){
+        $userEnteredAppInsightsInstrumentationKey = Read-Host  "Enter Application Insights instrumentation key or hit enter to accept the default [$appInsightsInstrumentationKey]"
+
+        if(![string]::IsNullOrEmpty($userEnteredAppInsightsInstrumentationKey)){   
+            $appInsightsInstrumentationKey = $userEnteredAppInsightsInstrumentationKey
+        }
+    }
+    if($appInsightsInstrumentationKey){ Add-InstallationSetting "identity" "appInsightsInstrumentationKey" "$appInsightsInstrumentationKey" | Out-Null }
+    if($appInsightsInstrumentationKey){ Add-InstallationSetting "common" "appInsightsInstrumentationKey" "$appInsightsInstrumentationKey" | Out-Null }
+    return $appInsightsInstrumentationKey
+}
+
+function Get-SqlServerAddress([string] $sqlServerAddress, [bool] $quiet){
```

Comment:
should we default to current machine if no sql server is specified.  still should prompt user.

Reply:
I believe we set localhost as the default in the install.config, and we go with that if nothing is entered.